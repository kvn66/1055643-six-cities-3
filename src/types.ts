export type LocationType = {
  latitude: number;
  longitude: number;
  zoom: number;
}

export type CityType = {
  location: LocationType;
  name: string;
}

export type HostType = {
  id: number;
  name: string;
  avatarUrl: string;
  isPro: boolean;
}

export type CardType = {
  id: number;
  city: CityType;
  bedrooms: number;
  images: string[];
  description: string;
  goods: string[];
  host: HostType;
  isFavorite: boolean;
  isPremium: boolean;
  location: LocationType;
  maxAdults: number;
  previewImage: string;
  price: number;
  rating: number;
  title: string;
  type: string;
}

export type FavoritesForCityType = {
  city: string;
  cards: CardType[];
}

export type UserInfoType = {
  id: number;
  name: string;
  email: string;
  avatarUrl: string;
  isPro: boolean;
}

export type ClassNameType = {
  ARTICLE: string;
  IMAGE: string;
}

export type ReviewType = {
  id: number;
  comment: string;
  date: string;
  rating: number;
  user: HostType;
}

export type CommentPostType = {
  rating: number;
  comment: string;
};

export type AuthDataType = {
  email: string;
  password: string;
};

