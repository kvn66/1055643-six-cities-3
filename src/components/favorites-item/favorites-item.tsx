import * as React from "react";
import PropTypes from 'prop-types';
import {Link} from "react-router-dom";
import {MemoizedOfferSmallCard} from "../offer-small-card/offer-small-card.jsx";
import {AppRoute, CardClassName} from "../../const";

const FavoritesItem = (props) => {
  const {favoritesForCity} = props;
  const {city, cards} = favoritesForCity;
  const cardsList = cards.map((card) =>
    <MemoizedOfferSmallCard key={card.id} card={card} className={CardClassName.FAVORITES} />
  );

  return (
    <li className="favorites__locations-items">
      <div className="favorites__locations locations locations--current">
        <div className="locations__item">
          <Link className="locations__item-link" to={AppRoute.ROOT}>
            <span>{city}</span>
          </Link>
        </div>
      </div>
      <div className="favorites__places">
        {cardsList}
      </div>
    </li>
  );
};

FavoritesItem.propTypes = {
  favoritesForCity: PropTypes.exact({
    city: PropTypes.string.isRequired,
    cards: PropTypes.arrayOf(
        PropTypes.exact({
          id: PropTypes.number.isRequired,
          city: PropTypes.exact({
            location: PropTypes.exact({
              latitude: PropTypes.number.isRequired,
              longitude: PropTypes.number.isRequired,
              zoom: PropTypes.number.isRequired,
            }).isRequired,
            name: PropTypes.string.isRequired,
          }).isRequired,
          bedrooms: PropTypes.number.isRequired,
          images: PropTypes.array.isRequired,
          description: PropTypes.string.isRequired,
          goods: PropTypes.array.isRequired,
          host: PropTypes.exact({
            id: PropTypes.number.isRequired,
            name: PropTypes.string.isRequired,
            avatarUrl: PropTypes.string.isRequired,
            isPro: PropTypes.bool.isRequired
          }).isRequired,
          isFavorite: PropTypes.bool.isRequired,
          isPremium: PropTypes.bool.isRequired,
          location: PropTypes.exact({
            latitude: PropTypes.number.isRequired,
            longitude: PropTypes.number.isRequired,
            zoom: PropTypes.number.isRequired,
          }).isRequired,
          maxAdults: PropTypes.number.isRequired,
          previewImage: PropTypes.string.isRequired,
          price: PropTypes.number.isRequired,
          rating: PropTypes.number.isRequired,
          title: PropTypes.string.isRequired,
          type: PropTypes.string.isRequired,
        }).isRequired
    ).isRequired,
  }).isRequired,
};

export const MemoizedFavoritesItem = React.memo(FavoritesItem);

export default FavoritesItem;
