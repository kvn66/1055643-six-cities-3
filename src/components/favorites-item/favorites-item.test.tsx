import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {MemoizedFavoritesItem} from "./favorites-item";
import {BrowserRouter} from "react-router-dom";
import configureMockStore from "redux-mock-store";
import {Provider} from "react-redux";
import {FavoritesForCityType} from "../../types";

const mockStore = configureMockStore();

const favoritesForCity: FavoritesForCityType = {
  city: `Amsterdam`,
  cards: [
    {
      id: 0,
      city: {
        location: {
          latitude: 52.38333,
          longitude: 4.9,
          zoom: 12,
        },
        name: `Amsterdam`,
      },
      bedrooms: 2,
      images: [`/img/apartment-01.jpg`, `/img/room.jpg`],
      description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      goods: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
      host: {
        id: 0,
        name: `Angelina`,
        avatarUrl: `/img/avatar-angelina.jpg`,
        isPro: true
      },
      isFavorite: true,
      isPremium: true,
      location: {
        latitude: 52.3909553943508,
        longitude: 4.85309666406198,
        zoom: 12,
      },
      maxAdults: 3,
      previewImage: `/img/apartment-01.jpg`,
      price: 120,
      rating: 4.5,
      title: `Beautiful &amp; luxurious apartment at great location`,
      type: `apartment`,
    },
    {
      id: 1,
      city: {
        location: {
          latitude: 52.38333,
          longitude: 4.9,
          zoom: 12,
        },
        name: `Amsterdam`,
      },
      bedrooms: 2,
      images: [`/img/room.jpg`, `/img/apartment-02.jpg`],
      description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
      goods: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
      host: {
        id: 0,
        name: `Angelina`,
        avatarUrl: `/img/avatar-angelina.jpg`,
        isPro: true
      },
      isFavorite: true,
      isPremium: true,
      location: {
        latitude: 52.369553943508,
        longitude: 4.85309666406198,
        zoom: 12,
      },
      maxAdults: 3,
      previewImage: `/img/room.jpg`,
      price: 120,
      rating: 4.5,
      title: `Beautiful &amp; luxurious apartment at great location`,
      type: `apartment`,
    }
  ]
};

it(`Render MemoizedFavoritesItem`, () => {

  const store = mockStore({});

  const tree = renderer
    .create(
        <BrowserRouter>
          <Provider store={store}>
            <MemoizedFavoritesItem
              favoritesForCity={favoritesForCity}
            />
          </Provider>
        </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
