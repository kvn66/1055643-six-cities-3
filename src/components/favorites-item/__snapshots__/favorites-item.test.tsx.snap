// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render MemoizedFavoritesItem 1`] = `
<li
  className="favorites__locations-items"
>
  <div
    className="favorites__locations locations locations--current"
  >
    <div
      className="locations__item"
    >
      <a
        className="locations__item-link"
        href="/"
        onClick={[Function]}
      >
        <span>
          Amsterdam
        </span>
      </a>
    </div>
  </div>
  <div
    className="favorites__places"
  >
    <article
      className="favorites__card place-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="place-card__mark"
      >
        <span>
          Premium
        </span>
      </div>
      <div
        className="favorites__image-wrapper place-card__image-wrapper"
      >
        <a
          href="/offer/0"
          onClick={[Function]}
        >
          <img
            alt="Place image"
            className="place-card__image"
            height="200"
            src="/img/apartment-01.jpg"
            width="260"
          />
        </a>
      </div>
      <div
        className="place-card__info"
      >
        <div
          className="place-card__price-wrapper"
        >
          <div
            className="place-card__price"
          >
            <b
              className="place-card__price-value"
            >
              €
              120
            </b>
            <span
              className="place-card__price-text"
            >
              / night
            </span>
          </div>
          <button
            className="place-card__bookmark-button place-card__bookmark-button--active button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="place-card__bookmark-icon"
              height="19"
              width="18"
            >
              <use
                xlinkHref="#icon-bookmark"
              />
            </svg>
            <span
              className="visually-hidden"
            >
              In bookmarks
            </span>
          </button>
        </div>
        <div
          className="place-card__rating rating"
        >
          <div
            className="place-card__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "90%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <h2
          className="place-card__name"
        >
          <a
            href="/offer/0"
            onClick={[Function]}
          >
            Beautiful &amp; luxurious apartment at great location
          </a>
        </h2>
        <p
          className="place-card__type"
        >
          Apartment
        </p>
      </div>
    </article>
    <article
      className="favorites__card place-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="place-card__mark"
      >
        <span>
          Premium
        </span>
      </div>
      <div
        className="favorites__image-wrapper place-card__image-wrapper"
      >
        <a
          href="/offer/1"
          onClick={[Function]}
        >
          <img
            alt="Place image"
            className="place-card__image"
            height="200"
            src="/img/room.jpg"
            width="260"
          />
        </a>
      </div>
      <div
        className="place-card__info"
      >
        <div
          className="place-card__price-wrapper"
        >
          <div
            className="place-card__price"
          >
            <b
              className="place-card__price-value"
            >
              €
              120
            </b>
            <span
              className="place-card__price-text"
            >
              / night
            </span>
          </div>
          <button
            className="place-card__bookmark-button place-card__bookmark-button--active button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="place-card__bookmark-icon"
              height="19"
              width="18"
            >
              <use
                xlinkHref="#icon-bookmark"
              />
            </svg>
            <span
              className="visually-hidden"
            >
              In bookmarks
            </span>
          </button>
        </div>
        <div
          className="place-card__rating rating"
        >
          <div
            className="place-card__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "90%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <h2
          className="place-card__name"
        >
          <a
            href="/offer/1"
            onClick={[Function]}
          >
            Beautiful &amp; luxurious apartment at great location
          </a>
        </h2>
        <p
          className="place-card__type"
        >
          Apartment
        </p>
      </div>
    </article>
  </div>
</li>
`;
